crates/iceberg/src/writer/file_writer/parquet_writer.rs

```diff
@@ -0,0 +1,390 @@
+                let mut per_col_null_val_num: HashMap<i32, u64> = HashMap::new();
+                metadata.row_groups.iter().for_each(|group| {
+                    group.columns.iter().zip(field_ids.iter()).for_each(
+                        |(column_chunk, &field_id)| {
+                            if let Some(column_chunk_metadata) = &column_chunk.meta_data {
+                                *per_col_size.entry(field_id).or_insert(0) +=
+                                    column_chunk_metadata.total_compressed_size as u64;
+                                *per_col_val_num.entry(field_id).or_insert(0) +=
+                                    column_chunk_metadata.num_values as u64;
+                                *per_col_null_val_num.entry(field_id).or_insert(0_u64) +=
+                                    column_chunk_metadata
+                                        .statistics
+                                        .as_ref()
+                                        .map(|s| s.null_count)
+                                        .unwrap_or(None)
+                                        .unwrap_or(0) as u64;
+                            }
+                        },
+                    )
+                });
+                (per_col_size, per_col_val_num, per_col_null_val_num)
+            };
+
+        let mut builder = DataFileBuilder::default();
+        builder
+            .file_path(file_path)
+            .file_format(DataFileFormat::Parquet)
+            .record_count(metadata.num_rows as u64)
+            .file_size_in_bytes(written_size as u64)
+            .column_sizes(column_sizes)
+            .value_counts(value_counts)
+            .null_value_counts(null_value_counts)
+            // # TODO
+            // - nan_value_counts
+            // - lower_bounds
+            // - upper_bounds
+            .key_metadata(metadata.footer_signing_key_metadata.unwrap_or_default())
+            .split_offsets(
+                metadata
+                    .row_groups
+                    .iter()
+                    .filter_map(|group| group.file_offset)
+                    .collect(),
+            );
+        Ok(builder)
+    }
+}
+
+impl FileWriter for ParquetWriter {
+    async fn write(&mut self, batch: &arrow_array::RecordBatch) -> crate::Result<()> {
+        self.current_row_num += batch.num_rows();
+        self.writer.write(batch).await.map_err(|err| {
+            Error::new(
+                crate::ErrorKind::Unexpected,
+                "write error from parquet writer",
+            )
+            .with_source(err)
+        })?;
+        Ok(())
+    }
+
+    async fn close(self) -> crate::Result<Vec<crate::spec::DataFileBuilder>> {
+        let metadata = self.writer.close().await.map_err(|err| {
+            Error::new(
+                crate::ErrorKind::Unexpected,
+                "close error from parquet writer",
+            )
+            .with_source(err)
+        })?;
+        if self.current_row_num == 0 {
```

Comment:
We should move this to first part?

Reply:
We should close the writer first. Otherwise, the writer's behavior is undefined.